### PR TITLE
sql: add string_to_array builtin

### LINKED ISSUE #15042: sql: implement string_to_array builtin
Add support for the [`string_to_array` PostgreSQL builtin](https://www.postgresql.org/docs/9.6/static/functions-array.html), which splits a string on a supplied delimiter and returns an array of the results.

### PR DESCRIPTION
**Pull request by justinj**
Fixes #15042.

Release note (sql change): Added string_to_array builtin.

### CODE REVIEW

**Comment by jordanlewis** [comment]
:lgtm: what if someone set the delimiter and the null arg to the same thing? or to empty strings? any other tricky cases you can think of?

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22391#-:-L4bfMnuXFgrMI6ZScYi:b-elu70q)*

**Comment by justinj** [comment]
You were right, there was some lurking complexity in the implementation here that I didn't pick up on - PTAL

File: pkg/sql/sem/builtins/builtins.go [hunk 1]
```diff
@@ -93,6 +93,8 @@ func categorizeType(t types.T) string {
 
 var digitNames = []string{"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
 
+var emptyStr = ""
```
Comment (jordanlewis) [inline pkg/sql/sem/builtins/builtins.go]: I don't think this is used anymore?

File: pkg/sql/logictest/testdata/logic_test/array [hunk 2]
```diff
@@ -1016,3 +1016,63 @@ query T
...
+
+query T
+SELECT STRING_TO_ARRAY('a', 'foo', NULL)
+----
+{"a"}
+
+query T
+SELECT STRING_TO_ARRAY('foofoofoofoo', 'foo', 'foo')
+----
+{"","","","",""}
```
Comment (jordanlewis) [inline pkg/sql/logictest/testdata/logic_test/array]: like we talked about offline, need test cases (and probably matching functionality :) for the empty string delimiter, which is supposed to just return the entire input as a single value in an array.

**Comment by justinj** [comment]
PTAL!

**Comment by jordanlewis** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22391#-:-L5GhqlaaJH4V0akPLZa:bnfp4nl)*

**Comment by justinj** [comment]
TFTR!

**Pull request merged**